### PR TITLE
Auth: Handle when access token has already been refreshed in OAuth token sync

### PR DESCRIPTION
**Pull request by mgyongyosi**
**What is this feature?**
Improves the way the OAuth token sync hook works by reducing the number of error messages and handling the case when the access token is already refreshed by another request/node.

**Why do we need this feature?**

**Who is this feature for?**

**Which issue(s) does this PR fix?**:

Fixes #

**Special notes for your reviewer:**

Please check that:
- [x] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

File: pkg/services/authn/authnimpl/sync/oauth_token_sync.go [hunk 1]
```diff
@@ -100,27 +103,49 @@ func (s *OAuthTokenSync) SyncOauthTokenHook(ctx context.Context, identity *authn
...
-	defer cancel()
 
-	if err := s.service.TryTokenRefresh(updateCtx, token); err != nil {
-		if errors.Is(err, context.Canceled) {
-			return nil
-		}
-		if !errors.Is(err, oauthtoken.ErrNoRefreshTokenFound) {
-			s.log.Error("Failed to refresh OAuth access token", "id", identity.ID, "error", err)
-		}
+	lockKey := fmt.Sprintf("oauth-token-sync-%s", identity.ID)
```
Comment (kalleep) [inline pkg/services/authn/authnimpl/sync/oauth_token_sync.go:107]: NIT: identity.ID should be unique here so we don't have to prefix the key

File: pkg/services/authn/authnimpl/sync/oauth_token_sync.go [hunk 2]
```diff
@@ -100,27 +103,49 @@ func (s *OAuthTokenSync) SyncOauthTokenHook(ctx context.Context, identity *authn
...
+				return nil, err
+			}
 
-		if err := s.sessionService.RevokeToken(ctx, identity.SessionToken, false); err != nil {
-			s.log.Warn("Failed to revoke session token", "id", identity.ID, "tokenId", identity.SessionToken.Id, "error", err)
+			// if the access token has already been refreshed by another request (for example in HA scenario)
+			tokenExpires := token.OAuthExpiry.Round(0).Add(-oauthtoken.ExpiryDelta)
+			if !tokenExpires.Before(time.Now()) {
+				return nil, nil
+			}
```
Comment (kalleep) [inline pkg/services/authn/authnimpl/sync/oauth_token_sync.go:130]: This seems like a repeat of what we are doing earlier in the function, but there we are checking id token also which I guess is not really needed here. But could still be a good idea to extract this to a function. wdyt?

File: pkg/services/authn/authnimpl/sync/oauth_token_sync.go (hunk 1 above)
Comment (mgyongyosi) [reply pkg/services/authn/authnimpl/sync/oauth_token_sync.go]: Changed, thanks!

File: pkg/services/authn/authnimpl/sync/oauth_token_sync.go (hunk 2 above)
Comment (mgyongyosi) [reply pkg/services/authn/authnimpl/sync/oauth_token_sync.go]: Yeah, maybe extracting the adjustment for the clock skew and expiry check make sense. I extracted those into the `getExpiryWithSkew` function.

**Review by kalleep — APPROVED** [review]
LGTM, is there a way to add test for this, or is that not possible with current test setup?

**Comment by mgyongyosi** [comment]
> LGTM, is there a way to add test for this, or is that not possible with current test setup?

I don't think it's possible. At least not easily. But I'm gonna spend some time to find a way to test this.

**Pull request merged**